### DISCUSSION TITLE
Problem with generating betatron radiation and followed by visualization (analysis and plotting) of the radiation spectra

### DISCUSSION THREAD

**Discussion by ddghosal** [BLAST-WarpX/warpx#4724] [Data Analysis]
I am trying to calculate betatron radiation produced by an electron beam and further to plot the radiation spectra for visualization. But the plot in the end looks weird. I tried the both option as Intensity Vs energy and frequency.

Instead of pasting the code here directly, I attached the two files. The ".py" extension won't be possible to attach, hence I made them as a txt file. Please pardon me for that.

[pywarpx_e_test_BR.txt](https://github.com/ECP-WarpX/WarpX/files/14384346/pywarpx_e_test_BR.txt)
[betatronspectra_plotting.txt](https://github.com/ECP-WarpX/WarpX/files/14384349/betatronspectra_plotting.txt)

The 1st attachment (i.e. pywarpx_e_test_BR.txt) is the pywarpx code(below) that I am using to generate betatron radiation.
Then, the 2nd attachment (i.e. betatronspectra_plotting.txt) is regarding the analysis and visualization of the spectra.

In the end, I need to use the betatron radiation dataset to predict beam spot size with the ML model. But, the plotted betatron radiation spectra (I tried both in terms of energy and frequency) looks weird. Can anyone please help me pointing out the mistakes I might be doing? or help me modifying the code if necessary? I'm definitely missing something!
Many thanks!

**Comment by RemiLehe** [comment]
Thanks for the question.
From looking at `betatronspectra_plotting.txt`, it seems that the way you are calculating betatron radiation is not quite correct. In `betatronspectra_plotting.txt`, it seems that you are extracting the momenta of the electron beam and histogramming the corresponding energy, and that you are considering this as the betatron radiation. I don't think that this is what betatron radiation is (instead, the quantity that you are plotting is simply the energy spectrum of the electron beam).

In order to obtain betatron radiation, you could try to use the quantum synchrotron module in the WarpX simulation (see `do_qed_quantum_synchrotron` [here](https://warpx.readthedocs.io/en/latest/usage/parameters.html)), which will result in the creation of photon particles during the simulation. By histogramming the properties of these photons, you should be able to get a correct estimate of the betatron spectrum.

**Reply by ddghosal** [reply]
Many thanks @RemiLehe 
I am following your link and the documenation on that page, struggling a bit at the moment though! I don't come with a good PIC background, so implementing the quantum synchrotron module in my PyWarpX code might be challenging for me, but nevertheless I will surely give it a try. Just to be sure- this feature requires QED process (such as beamstrahlung etc.) to consider, right? can it be used to calculate classical Betatron radiation also (for example my case)?
If so, could you please help me with some coding idea, how to integrate this within my existing pywarpx code (that I shared earlier)?

Secondly, I am trying to explore the openCL and PyOpenCL frameworks and trying to use SynchRad if that can help me with the radiation creation and analysis, plotting etc.
[url](https://github.com/hightower8083/synchrad)
best,
